File: python/WebClient.py

import urllib2

baseUrl = "http://localhost:8080"

myToken = ""
permChallenge = ""
tempChallenge = ""

def connect():
    global myToken
    url = "{0}/newClient".format(baseUrl)
    myToken = urllib2.urlopen(url).read()
    

def requestChallenge(uid,perm_pk,temp_pk):
    global permChallenge, tempChallenge
    print "Requesting challenge."
    url = "{0}/request_challenge?uid={1}&perm_pk={2}&temp_pk={3}&token={4}".format(baseUrl,uid,perm_pk,temp_pk,myToken)
    res = urllib2.urlopen(url).read()
    print "FROM SERVER".format(res)
    if (',' in res):
        (permChallenge,tempChallenge) = res.split(',')
    else:
        permChallenge = ""
        tempChallenge = ""

def sendResponse(uid,perm_response,temp_response):
    print "Sending response."
    url = "{0}/response?uid={1}&perm_response={2}&temp_response={3}&token={4}".format(baseUrl,uid,perm_response,temp_response,myToken)
    res = urllib2.urlopen(url).read()
    print "FROM SERVER: {0}".format(res)

def logout():
    print "Logging out"
    url = "{0}/shoeDisconnected?token={1}".format(baseUrl,myToken)
    res = urllib2.urlopen(url).read()
    print "FROM SERVER: {0}".format(res)

if __name__=="__main__":
    connect()
    print "{0}/newSession/{1}".format(baseUrl,myToken)
    print "commands: requestChallenge sendResponse logout"
    while 1:
        cmd = raw_input()
        argv = cmd.split()
        if (argv[0]=="requestChallenge"):
            if (len(argv)==4):
                requestChallenge(argv[1],argv[2],argv[3])
            else:
                print "Usage: requestChallenge uid perm_pk temp_pk"
        elif (argv[0]=="sendResponse"):
            if (len(argv)==4):
                sendResponse(argv[1],argv[2],argv[3])
            else:
                print "Usage: sendResponse uid perm_response temp_response"
        elif (argv[0]=="logout"):
            if (len(argv)==1):
                logout()
            else:
                print "Usage: logout"

File: native_client/WebClient.java
import java.io.*;
import java.net.*;

public class WebClient {

    private String baseUrl; //ends with /

    private String myToken;
    private String permChallenge;
    private String tempChallenge;

    public WebClient(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    /////////////
    // Getters //
    /////////////

    public String getMyToken() {
        return myToken;
    }

    public String getPermChallenge() {
        return permChallenge;
    }

    public String getTempChallenge() {
        return tempChallenge;
    }

    ///////////////////////////////
    // Interface with web server //
    ///////////////////////////////

    // Get a unique id from the server
    public void connect() {
        System.out.println("Checking with server for unique token.");

        try {
          StringBuilder result = new StringBuilder();
          URL url = new URL(this.baseUrl + "newClient");
          HttpURLConnection conn = (HttpURLConnection) url.openConnection();
          conn.setRequestMethod("GET");

          BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
          String line;
          while ((line = rd.readLine()) != null) {
             result.append(line);
          }
          rd.close();
          //System.out.println("localhost:8080/newSession/" + result.toString());
          this.myToken = result.toString();
          System.out.println("Token received from server " + this.myToken);
        } catch (Exception e) {
          System.out.println(e);
        }
    }

    // Request a challenge from the server
    public void requestChallenge(String uid, String perm_pk, String temp_pk) {
        System.out.println("Requesting challenge");

        try {
          StringBuilder result = new StringBuilder();
          String urlString =  new String(
              this.baseUrl + "request_challenge"+
              "?uid="+uid+
              "&perm_pk="+perm_pk+
              "&temp_pk="+temp_pk+
              "&token="+this.myToken);

          URL url = new URL(urlString);
          HttpURLConnection conn = (HttpURLConnection) url.openConnection();
          conn.setRequestMethod("GET");

          BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
          String line;
          while ((line = rd.readLine()) != null) {
             result.append(line);
          }
          rd.close();
          System.out.println(result.toString());
        } catch (Exception e) {
          System.out.println(e);
        }
    }

    // Send the response to the challenge from the server
    public void sendResponse(String uid, String perm_response, String temp_response) {
        System.out.println("Sending response.");

        try {
          StringBuilder result = new StringBuilder();
          String urlString =  new String(
              this.baseUrl + "response"+
              "?uid="+uid+
              "&perm_response="+perm_response+
              "&temp_response="+temp_response+
              "&token="+this.myToken);

          URL url = new URL(urlString);
          HttpURLConnection conn = (HttpURLConnection) url.openConnection();
          conn.setRequestMethod("POST");

          BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
          String line;
          while ((line = rd.readLine()) != null) {
             result.append(line);
          }
          rd.close();
          System.out.println(result.toString());
        } catch (Exception e) {
          System.out.println(e);
        }


    }

    // Log the user out when the device is disconnected
    public void logout() {
        System.out.println("Logging user out.");
        
        try {
          StringBuilder result = new StringBuilder();
          String urlString =  new String(
              this.baseUrl + "shoeDisconnected"+
              "&token="+this.myToken);

          URL url = new URL(urlString);
          HttpURLConnection conn = (HttpURLConnection) url.openConnection();
          conn.setRequestMethod("POST");

          BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
          String line;
          while ((line = rd.readLine()) != null) {
             result.append(line);
          }
          rd.close();
          System.out.println(result.toString());
        } catch (Exception e) {
          System.out.println(e);
        }
    }
}
